File: js/Chess/Piece.js
import pieceColor from './PieceColor'

export default class Piece {
    constructor(pieceType, color) {
        this.pieceType = pieceType;
        this.color = color;
    }

    getMoves(location, board) {
        return this.pieceType.getMoves(location, board);
    }

    toString() {
        const c = this.pieceType === null ? ' .' : ' ' + this.pieceType;
        return this.color = pieceColor.black ? c.toLowerCase() : c;
    }

    is(color, pieceType) {
        const isColorEqual = this.color === color;
        if (pieceType) {
            return isColorEqual && this.pieceType === pieceType;
        } else {
            return isColorEqual;
        }
    }
}

File: cs/Chess/Piece.cs
using System.Collections.Generic;

namespace Chess
{
    public class Piece
    {
        public readonly PieceColor Color;
        private readonly PieceType pieceType;

        public Piece(PieceType pieceType, PieceColor color)
        {
            this.pieceType = pieceType;
            Color = color;
        }

        public IEnumerable<Location> GetMoves(Location location, Board board) => 
            pieceType.GetMoves(location, board);

        public override string ToString()
        {
            var c = pieceType == null 
                ? " ." 
                : $" {pieceType}";
            
            return Color == PieceColor.Black 
                ? c.ToLower() 
                : c;
        }

        public static bool Is(Piece piece, PieceColor color) =>
            piece != null && piece.Color == color;

        public static bool Is(Piece piece, PieceColor color, PieceType pieceType) =>
            Is(piece, color) && piece.pieceType == pieceType;
    }
}